File: Xenon_VSCode/Homer.ts
import CAlien = require("Alien");
import CHomerProjectileWeapon = require("HomerProjectileWeapon");
import gsCControls = require("Controls");
import gsCTimer = require("Timer");
import enums = require("Enums");
import CShip = require("Ship");
import gsCVector = require("Vector");
import CPlayGameState = require("PlayGameState");
import CExplode = require("Exploder");

class CHomer extends CAlien {
    HOMER_MAX_XSPEED: number = 3.0;
    HOMER_XSPEED_SCALE: number = 0.025;

    constructor(playGameState: CPlayGameState) {
        super();
        this.m_playGameState = playGameState;
        this.m_name = "Homer";
    }

    //-------------------------------------------------------------

    public getActorInfo() {
        this.m_actorInfo = this.m_scene.GetlistOfActors();
        return this.m_actorInfo.GetActorInfoListItem(enums.ActorInfoType.INFO_HOMER);
    }

    //-------------------------------------------------------------

    public activate(): boolean {
        return super.activate();
    }

    //-------------------------------------------------------------

    public update(controls: gsCControls, gameTime: gsCTimer): boolean {

        if (this.m_shield == 0) {
            var weapon: CHomerProjectileWeapon = new CHomerProjectileWeapon(this.m_playGameState);
            this.m_scene.addActor(weapon);
            weapon.activate();
            weapon.setOwner(this);
            weapon.setOffset(new gsCVector(0.0, 0.0));
            weapon.detonate();
            var explode = new CExplode(this);
            super.kill();
            return true;
        }

        var ship: CShip = this.m_scene.findShip();
        var dx: number = 0.0;

        if (ship != null) {
            dx = ship.getPosition().X - this.m_position.X;
        }
        dx *= this.HOMER_XSPEED_SCALE;

        if (dx < -this.HOMER_MAX_XSPEED) {
            dx = -this.HOMER_MAX_XSPEED;
        }
        if (dx > this.HOMER_MAX_XSPEED) {
            dx = this.HOMER_MAX_XSPEED;
        }

        //this.m_position += new gsCVector(dx, this.m_velocity.Y);
        this.m_position.plusEquals(new gsCVector(dx, this.m_velocity.Y));
        super.animate(enums.AnimationMode.ANIMATE_LOOP);

        return true;
    }

    //-------------------------------------------------------------
}

export = CHomer;

File: Xenon-Original_C++_Code/xenon/source/homer.cpp
//-------------------------------------------------------------
//
// Class:	CHomer
//
// Author:	John M Phillips
//
// Started:	06/05/00
//
// Base:	CAlien
//
// Derived:	None
//
//-------------------------------------------------------------

#include "game.h"

//-------------------------------------------------------------

CHomer::CHomer()
{
}

//-------------------------------------------------------------

CHomer::~CHomer()
{
}

//-------------------------------------------------------------

bool CHomer::activate()
{
	if (!isActive())
		m_timer.start();

	return CActor::activate();
}

//-------------------------------------------------------------

bool CHomer::update(Controls *controls)
{
	if (m_shield == 0) {

		CHomerProjectileWeapon *weapon = new CHomerProjectileWeapon;
		m_scene->addActor(weapon);
		weapon->activate();
		weapon->setOwner(this);
		weapon->setOffset(gsCVector(0.f,0.f));
		weapon->detonate();

		explode();
		kill();
		return true;
		}

	CShip *ship = m_scene->findShip();

	float dx = 0.f;

	if (ship)
		dx = ship->getPosition().getX() - m_position.getX();

	dx *= HOMER_XSPEED_SCALE;

	if (dx < -HOMER_MAX_XSPEED)
		dx = -HOMER_MAX_XSPEED;
	if (dx > HOMER_MAX_XSPEED)
		dx = HOMER_MAX_XSPEED;

	m_position += gsCVector(dx,m_velocity.getY());;

	animate(ANIMATE_LOOP);
	
	return true;
}

//-------------------------------------------------------------
